`Tabs`: tweak sizing and overflow behavior of TabList

DaniGuardiola opened the pull request:
<!-- Thanks for contributing to Gutenberg! Please follow the Gutenberg Contributing Guidelines:
https://github.com/WordPress/gutenberg/blob/trunk/CONTRIBUTING.md -->

## What?
<!-- In a few words, what is the PR actually doing? -->

This PR modifies the behavior of `Tabs.TabList` in regards to sizing and overflow to address some needs.

## Why?
<!-- Why is this PR necessary? What problem is it solving? Reference any existing previous issue(s) or PR(s), but please add a short summary here, too -->

We have needs around proper management of overflowing tabs (https://github.com/WordPress/gutenberg/issues/64093) and tab lists that expand to fill their container (https://github.com/WordPress/gutenberg/pull/61974#issuecomment-2271217124).

## How?
<!-- How is your PR addressing the issue at hand? What are the implementation details? -->

- It makes the `TabList` have a default `width` of `fit-content` (only for horizontal orientation). This means that tabs will only take up the space needed for their content, and the tab list (container) will adjust its width to the space needed for the sum of its child tabs.
  - This is set with `:where` for reduced specificity, in order to make overrides easier.
- It sets `overflow-x` to `auto` in the `TabList`, making it overflow horizontally only when necessary.
- It makes tabs expand (`flex-grow: 1`) to fill up **additional** available space when the container (`TabList`) has it. Note that due to `TabList` having `width: fit-content` by default, this won't happen unless the user intentionally modifies the width, for example by setting `width: 100%` to fill up its container's size.
- It centers the contents of a tab horizontally (`justify-content: center`, only for horizontal orientation).

The TL;DR is:

- Tabs will scroll when necessary.
- Tabs hug contents by default.
- If a "fill" style is wanted, just set the `TabList` width you want and tabs will adapt gracefully.

There is an additional detail that is important now that the tab list scrolls: when a tab is selected, it is programmatically scrolled into view.

Finally, I've gone through every single instance of `Tabs` and applied any necessary tweaks to retain the same styles. Everything should look exactly the same after these changes.

## Testing Instructions
<!-- Please include step by step instructions on how to test this PR. -->
<!-- 1. Open a post or page. -->
<!-- 2. Insert a heading block. -->
<!-- 3. etc. -->

Open the storybook locally (`npm run storybook:dev`) and go to the "Tabs > Size And Overflow Playground" story. Play with it. Also, check that none of the instances of this component in Gutenberg are affected (see below).

---

Where to find all `Tabs` instances for testing:

1. This one requires a small tweak in the code, replace `canChooseAColor && canChooseAGradient` with `true` in `gutenberg/packages/block-editor/src/components/colors-gradients/control.js`, then open post editor > select some text > go to "Block" in right sidebar > click "Background" (it'll be the "second" tablist below the first one)
2. Open post editor > select some text > go to "Block" in right sidebar > click "Background"
3. Open post editor > open block inserter > check inside "Patterns" and "Media" (vertical tabs)
4. Open post editor > add a "navigation block" and select it > go to "Block" in the right sidebar (tabs with icon labels)
5. Open post editor > open block inserter and document overview
6. (With a "block" theme set like Twenty Twenty-Four) Appearance > Editor > Styles > Edit styles (pencil icon button) > Select a palette
7. (With a "block" theme set like Twenty Twenty-Four) Appearance > Editor > Styles > Edit styles (pencil icon button) > Typography > Manage fonts
8. (With a "block" theme set like Twenty Twenty-Four) Appearance > Editor > Styles > Edit styles (pencil icon button) > Style Book (eye icon button)
9. (With a "non-block" theme set like Twenty Fifteen) Appearance > Widgets (tabs i right sidebar)
10. Open post editor > open right sidebar
11. Open post editor > select some text > open "more" dropdown in the floating menu > Highlight
12. Open post editor > open options dropdown menu (vertical dots icon button) > Preferences (vertical tabs)

### Testing Instructions for Keyboard
<!-- How can you test the changes by using the keyboard only? Please note, this is required for PRs that change the user interface (UI). This ensures the PR can be tested for any possible accessibility regressions. -->

In the same story, you can tab into the tab list and use arrows to move.

## Screenshots or screencast <!-- if applicable -->

https://github.com/user-attachments/assets/ac2bdac8-bc7a-437e-8b62-de5dd3d1360b

DaniGuardiola committed:
Tweak sizing and overflow behavior of TabList.

DaniGuardiola committed:
Add size and overflow playground story.

DaniGuardiola committed:
Add "scroll into view" behavior to selected tabs.

DaniGuardiola committed:
fit-content only on horizontal orientation

jasmussen commented:
Have a look also at the `custom-scrollbars-on-hover` mixin, it may be a good fit here.

DaniGuardiola committed:
Reduce specificity of `fit-content` to make it easier to override.

DaniGuardiola committed:
centered label only when orientation is horizontal

DaniGuardiola committed:
Remove unused file.

DaniGuardiola committed:
Fix inspector controls tabs.

DaniGuardiola committed:
Fix font library modal tabs.

DaniGuardiola committed:
Fix style-book tabs.

DaniGuardiola committed:
fix typo

DaniGuardiola commented:
I checked every instance, but I couldn't find these two in the product, I'd appreciate help finding them so I can test them:

- https://github.com/WordPress/gutenberg/blob/9979d58d56a86270ac037b6d677b1a0d41b97463/packages/block-editor/src/components/colors-gradients/control.js#L138
- https://github.com/WordPress/gutenberg/blob/bf14552fcbf10a51aee015d459b6dcb5bf8420cd/packages/edit-widgets/src/components/sidebar/index.js#L44

DaniGuardiola commented:
@jasmussen that's SASS-only, which we don't have access to from this component. This would be easier to use if implemented in vanilla CSS, e.g. a utility class that uses CSS custom properties (A.K.A. CSS variables) as parameters.

DaniGuardiola committed:
Add changelog entry.

DaniGuardiola committed:
fix emotion being weird ugh

DaniGuardiola commented:
Thanks @tyxla, I was able to find them and check them. No changes needed, so I've now reviewed all uses and all should be correct.

Please give feedback when you can @WordPress/gutenberg-design. I really like this solution for its simplicity, though it might be good to find a way to indicate overflow more clearly, specially in mac where scrollbars are hidden by default until scroll is initiated. Here are some examples I found, sharing as inspiration:


https://github.com/user-attachments/assets/64b06223-2cfc-4ee0-b0ba-9337ba1baf4d

jameskoster commented:
The main question for me is; with horizontal scrolling do we want/need to continue wrapping tab labels?

Despite not as many tabs being in view initially, I prefer not wrapping. Wrapping text in such a narrow space is hard to read imo, though I appreciate this may be subjective.

| Wrapping | Not wrapping |
| --- | --- |
| <img width="586" alt="Screenshot 2024-08-09 at 14 18 47" src="https://github.com/user-attachments/assets/e3fa3467-051b-4a15-b1d8-be9723bf6f32"> | <img width="589" alt="Screenshot 2024-08-09 at 14 19 40" src="https://github.com/user-attachments/assets/9f78ff91-f3fa-4382-9a48-921c44c67aa3"> |

---

In terms of visually indicating the overflow, I'd lean towards a simple fade:

<img width="939" alt="ovefflow" src="https://github.com/user-attachments/assets/1aa62ebe-090b-4ccc-add3-71c62f5819fb">

Is there a way to do this so it works on any background color? 🤔 

An icon only seems useful to me if it's interactive (facilitating scrolling). I don't know if that's necessary though, probably one to explore separately with some a11y feedback. 

---

As Joen notes, it would be good to do something with the scrollbar styling, even if in a follow-up.

tyxla reviewed (commented):
Thanks for working on this one @DaniGuardiola! I appreciate the dedicated story, this is so helpful to have!

Could you please provide some testing instructions on testing in the editor?  We're touching a bunch of editor components, and we should ensure everything continues to look and work well.

DaniGuardiola commented:
@tyxla I guess you can modify an instance to add extra tabs or make existing tabs longer, either through code or using the inspector. Or, try to find a locale that causes the overflow for the tabs instance. You can also get creative and force smaller widths in containers and things like that. In the end it depends on the specific instance and its context.

tyxla commented:
> @tyxla I guess you can modify an instance to add extra tabs or make existing tabs longer, either through code or using the inspector. Or, try to find a locale that causes the overflow for the tabs instance. You can also get creative and force smaller widths in containers and things like that. In the end it depends on the specific instance and its context.

Oh yes, I get that, thanks. I meant more about adding specific instructions about the concrete components that are affected by all changes. I see a bunch of changes to the block editor, widget editor, and site editor, and I don't see testing instructions related to testing those instances. Hope that makes sense.

DaniGuardiola commented:
To be clear, you mean a complete list on where to find each instance of Tabs? Happy to do so, just want to double check before I do the work to write that up :P

This PR does not modify their appearance so in any case the testing would be to verify that none are different or broken now.

tyxla commented:
Yep, that's what I had in mind. Thanks in advance!

ciampo commented:
> The main question for me is; with horizontal scrolling do we want/need to continue wrapping tab labels?
> 
> Despite not as many tabs being in view initially, I prefer not wrapping. Wrapping text in such a narrow space is hard to read imo, though I appreciate this may be subjective.

Maybe we can default to "no wrapping" and offer an easy path to consumers of `Tab` to change that behavior (either via showing the CSS tweaks to apply, or by exposing a prop)


> I'd lean towards a simple fade
> [...]
> Is there a way to do this so it works on any background color? 🤔

IMO this is the simplest and most elegant approach, and it should be easy to implement — we could use [`mask-image`](https://developer.mozilla.org/en-US/docs/Web/CSS/mask-image) on the tablist element. Something like `mask-image: linear-gradient(to right, black, transparent);` would be a good starting point and should be background-color agnostic.

The tricky part is to make sure that the fade (and therefore the visual hint to scroll the tablist) is always perceivable, even the last visible tabs in the list perfectly aligns with the edge of the tablist.

jameskoster commented:
> Maybe we can default to "no wrapping" and offer an easy path to consumers of Tab to change that behavior (either via showing the CSS tweaks to apply, or by exposing a prop)

To be honest a prop feels like overkill to me. Wrapping should only be considered when the label is very long, and very long labels should be strongly discouraged. So I think we can consider it an edge case.

> even the last visible tabs in the list perfectly aligns with the edge of the tablist.

Yes this is a bit tricky. A secondary affordance (icon) may be necessary after all. But choosing the right icon is challenging... a chevron is the obvious choice but it kind of looks interactive. Additionally an icon may be confusing when the tabs themselves are icon based. 

<img width="939" alt="tabs" src="https://github.com/user-attachments/assets/9a417e5d-4e51-4198-985d-f863783dcb05">

I think this detail may be worth exploring deeply in a dedicated issue.

DaniGuardiola committed:
Merge branch 'trunk' of https://github.com/WordPress/gutenberg into feat/tablist-sizing-and-overflow-improvements

DaniGuardiola commented:
Hi folks, I'm back from vacation. It seems like there's agreement in this thread about the basics of this PR, and about working on further improvements (regarding signaling that there's overflow to the user) in follow-ups, so I'm putting this PR up for formal review now.

To facilitate testing (on @tyxla's suggestion), I added a list of every instance of `Tabs` in the description. Any of them could be affected, even the ones I didn't have to tweak, so please go over them and let me know if you spot any issues.

jameskoster commented:
> regarding signaling that there's overflow to the user

Can we add the fade in this PR? There's a more nuanced issue about indicating overflow when the fade isn't visible, but I think that's a detail to address separately.

The other decisions to make relates to wrapping. My feeling is that wrapping labels _and_ scrolling feels like overkill. I'd prefer to remove the wrapping and provide guidelines to avoid the use of extremely long tab labels.

ciampo commented on packages/edit-site/src/components/style-book/index.js:
```diff
@@ -249,16 +249,18 @@ function StyleBook( {
-									</Tabs.Tab>
-								) ) }
-							</Tabs.TabList>
+							<div className="edit-site-style-book__tablist-container">
```
Do we need this extra `div`, or could we assign styles to the tablist directly?

DaniGuardiola replied:
Feel free to try, and I don't remember the details now but I do remember arriving to the conclusion that it was necessary. Trust me, I always try to go the minimal route :) - I could've missed something though!

ciampo reviewed (commented):
I took a high-level look at the code changes as things are LGTM. If no-one else gets to it, I'll also take a look at the changes across the editor later.

If it's not too big of a task, I agree with @jameskoster that we should try to add at least a gradient for when items are overflowing (we could try with `mask-image` as suggested [here](https://github.com/WordPress/gutenberg/pull/64371#issuecomment-2288513179))

afercia commented:
As mentioned on https://github.com/WordPress/gutenberg/issues/64093#issuecomment-2324289015, horizontal scrolling is an usability and accessibility problem and should be avoided when possible. 
Also, scrollable elements need some special treatment, as the native browser behavior is different as mentioned on https://github.com/WordPress/gutenberg/issues/64093#issuecomment-2324289015, see links posted there for more details.

DaniGuardiola commented:
@afercia appreciate your thoughts, though I would highlight that (correct me if I'm wrong folks) this does not necessarily come with a recommendation to have long lists of tabs that overflow, but rather a way to cleanly address that edge case of overflowing tabs in a way that remains accessible and looks good. My understanding (again, correct me if I'm wrong @WordPress/gutenberg-design) is that the guidelines are to try and keep tabs short enough to prevent this situation in the first place. But when they do overflow, we want them to not be broken or look bad.

Regarding the accessibility problems you mention, would you mind being specific? Can you list under what circumstances and input methods (e.g. screen reader, mouse, keyboard, etc) this UI is not accessible? I'd like to understand these issues well so I can explore solutions or consider whether scroll is really not a good enough solution altogether.

tyxla reviewed (dismissed):
Something seems to be off with the vertical tabs inside Patterns in the block inserter:

![Screenshot 2024-09-02 at 16 54 53](https://github.com/user-attachments/assets/dae4dbc5-2976-4c1f-901a-13b2823d09db)

The preferences vertical tabs test well FWIW.

tyxla reviewed (commented):
BTW, thanks for taking the time to add thorough testing instructions @DaniGuardiola. 

Besides the one I mentioned above, I've verified the rest of the instances worked well 👍

afercia commented:
> Regarding the accessibility problems you mention, would you mind being specific? Can you list under what circumstances and input methods (e.g. screen reader, mouse, keyboard, etc) this UI is not accessible? I'd like to understand these issues well so I can explore solutions or consider whether scroll is really not a good enough solution altogether.

@DaniGuardiola sure, but it's a TL;DR. To me, these are the main concerns:

1
Browsers treat scrollable elements with overflow in a different way. Firefox is the only browser that makes scrollable elements _focusable_ (and only when there is actual content overflow). Other browsers don't. This introduces a problem because a tab stop on a non-labeled, unsemantic, element isn't OK for accessibility. Also, the user experience would be different depending on the browser in use. Steve Faulkner (co-author of multiple W3C specs) suggested years ago [a technique](https://www.tpgi.com/short-note-on-improving-usability-of-scrollable-regions/) to make this behavior consistent across browsers. The only point of that technique I'd consider a bit arguable is the ARIA role set on the scrollable div (`region`) that, together with the aria-label, makes that element be exposed as an [ARIA landmark](https://www.w3.org/WAI/ARIA/apg/patterns/landmarks/examples/general-principles.html). Instead, ARIA landmarks should be used to identigy the main sections of a page.

You may want to test the Storybook example with Firefox and see that the container with overflow is focusable. Screenshot:

![Screenshot 2024-09-05 at 09 52 49](https://github.com/user-attachments/assets/57942519-194e-4e2e-a1e1-a6a271fc95ec)

For more details on scrollable elements with overflow, see https://github.com/WordPress/gutenberg/issues/45809

See also the work done for the modal dialog scrollable content on https://github.com/WordPress/gutenberg/pull/47426. I would have loved to abstract that work and implement it for the Scrollable component but haven't found time and opportunity to do that.

2
Scrollbars. 
Modern operating systems do have user preferences to always show scrollbars. When enabled, the scrollable container will show scrollbars. When the setting is 'auto`, it will show scrollbars only when scrolling. See the screenshot above, where the scrollbars are visible. However, the shape and size of the scrollbars vary across operating systems and browsers. As such, we can't predict how much space should be reserved at the bottom of the tabs list to prevent the scrollbar from hiding content.

3
Usability and accessibility aren't just about keyboard and screen readers. Non-tech savvy users, users with cognitive impairments, users with motor impairments etc. will potentially struggle understand the content is horizontally scollable and actually interact withl it. 

4
Lastly, a little more philosophical consideration. The very nature of content on the web is to reflow and adapt to different devices and viewports. Instead of attempting to preserve a fixed horizontal layout, to me a better approach would be to just make the tabs _adapt_ and transform to different viewports and always keep all the content visible.

Overall, my personal opinion is that horizontal scrolling introduces more harms than good and should be avoided. I genuinely appreciate the effort put in this exploration but I do think an alternate solution should be explored.

DaniGuardiola committed:
Merge branch 'trunk' of https://github.com/WordPress/gutenberg into feat/tablist-sizing-and-overflow-improvements

DaniGuardiola committed:
Prevent unwanted focusable container in Firefox.

DaniGuardiola committed:
Add fade effect.

DaniGuardiola committed:
Merge branch 'trunk' of https://github.com/WordPress/gutenberg into feat/tablist-sizing-and-overflow-improvements

DaniGuardiola committed:
Fix IntersectionObserver logic.

DaniGuardiola committed:
Feature detect IntersectionObserver to prevent tests from failing.

DaniGuardiola commented:
@afercia 

1. Solved, container is not focusable now in any browser.
2. The way the tablist layout works allows for any and all scrollbar sizes so this won't be an issue.
3. I agree we should put effort into communicating clearly that there's overflow and facilitating interaction for users. The fade effect is a start, but we can go further in a follow-up, as @jameskoster was exploring.
4. It is not true that this does more harm than good. There seems to be consensus around a general recommendation to never let tabs overflow in the first place. This work is to cover the edge cases where it might happen anyway, to not leave it broken for everyone. It needs to work generically under any situation where the viewport is just not large enough.

Further explorations can be done in the future, I personally welcome any improvement, however this seems like a good and realistic approach that will fix a real issue for users right away. I simply do not have more time to dedicate to this issue and leaving tabs broken across the app is unacceptable, as it should be accessible to everyone. See @jameskoster example: https://github.com/WordPress/gutenberg/issues/64093#issuecomment-2337963915

I appreciate if you want to keep the discussion going to find a good alternative to this that will have less trade-offs, and I hope that someone with enough time and resources can help to iteratively explore those ideas. Me and my team simply do not have those resources at the moment, I hope you understand.

In the short term, I'd welcome any suggestions to improve the current approach.

@jameskoster can you check the current state of the PR? I implemented the fade effect. I think your idea to add chevrons is great to make the overflow obvious. I was thinking about this and about how to further improve usability for users. You mentioned that you'd expect these indicators to be interactive and I agree! Here's my idea:

- Add those indicators.
- They are NOT focusable, and hidden to screen readers. It's just redundant for keyboard and screen reader users: the tree contains all tabs for non-sighted users, and interaction is unchanged for keyboard users (just use left/right arrows).
-  They ARE interactive. Activating them with a pointer input (mouse, touch...) will scroll the tabs a bit.

Alternatively, we can make the tablist scroll at a fixed speed when hovering those arrows.

In any case, this will make it more useable and discoverable for all users, even when the scrollbar is hidden (Mac), without getting in the way of screen reader and keyboard users. Thoughts?

jameskoster commented:
I appreciate horizontal scrolling is not perfect, but imo it is better than a _broken component_. Since there is no clear and obvious alternative, and given that scrolling is the most common approach one finds in other design systems (suggesting it will provide a familiar pattern), I'd vote to merge this.

As Dani says we can revisit when a better technique is found.

@DaniGuardiola I think we can remove text wrapping on tab labels now. Also I'm seeing a glitch with the active tab indicator:


https://github.com/user-attachments/assets/3206cb3e-8402-400b-aaeb-ae91e6f2d2d2

> Here's my idea

I think the idea sounds good in principle. I presume that will be added in a follow-up?

tyxla commented:
I've also noticed that the vertical tab indicator is gone:


https://github.com/user-attachments/assets/1b820b8c-c193-4f91-8bee-13ce4ecb0ad5



as compared to `trunk`:


https://github.com/user-attachments/assets/5a251057-7d9f-4814-a0c7-92f63e1a3faa

DaniGuardiola committed:
Merge branch 'trunk' of https://github.com/WordPress/gutenberg into feat/tablist-sizing-and-overflow-improvements

DaniGuardiola committed:
Add a bit of tolerance for scroll state detection.

DaniGuardiola committed:
Fix vertical indicator.

DaniGuardiola committed:
Better handling of vertical overflow.

DaniGuardiola committed:
Add a bit of scroll margin for better "scroll into view" experience.

DaniGuardiola committed:
Horizontal fade should only happen on horizontal direction.

DaniGuardiola committed:
Adjust for offset parent scroll state in `getElementOffsetRect`.

DaniGuardiola committed:
Better "scroll into view" positioning heuristics ("nearest").

DaniGuardiola commented:
@jameskoster @tyxla fixed bugs:

- Wrong indicator position when scrolled (also works in RTL).
- Vertical indicator restored (it was a dumb typo).

Also:

- Added a bit of tolerance to the intersection observer to prevent an edge case (probably related to rounding) where it'd trigger a false positive even when there's no scroll.
- Polished horizontal scroll a bit, even though we don't really plan for it. See commit history for details.
- A few bits of polish around scrolling behavior, see commit history for details.

DaniGuardiola committed:
Invert use of before and after to remove z-index and fix related issues.

DaniGuardiola committed:
Make vertical indicator light blue as discussed.

DaniGuardiola committed:
Undo most overrides in pattern/media vertical tabs.

DaniGuardiola committed:
Clean up outdated styles previously needed for label wrapping.

DaniGuardiola commented:
@tyxla fixed patterns/media vertical tabs.

Related (cc @jameskoster @WordPress/gutenberg-design) I updated the vertical indicator to be light blue across all instances as discussed here: https://github.com/WordPress/gutenberg/pull/62879#issuecomment-2218239017

I removed most style overrides from the patterns/media vertical tabs (inserter) but retained the padding since it affects the spacing so it'd be a major difference. Still, there are a few differences with the previous styles (focus-visible ring, text weight...) so please do check it out. Ideally, we'd find a way to unify these styles too. Maybe we could have 1. a "narrow" variant for vertical tabs and 2. native support for trailing icons that keeps things balance.

This diff should offer additional hints in regards to the changes: https://github.com/WordPress/gutenberg/pull/64371/commits/1d064ae7e5362714f183c1f9609927dc6cd93050

One last thing, I realized that a previous change (adding flex-shrink 0) had the side effect of preventing wrapping, which was planned anyway, so I went ahead and did a bit of cleanup to remove outdated CSS properties that don't have any effect now.

So please help me test everything (I did some testing myself but appreciate if we can do a final pass together to make sure everything works as expected) and let's look into landing this :)

DaniGuardiola commented:
I gotta say, the light blue indicator seems to be a bit too light, I kinda preferred the grey one. Maybe the opacity (4%) is too low and we could use a higher value for better contrast. I struggle myself to see it and I'm sighted, so I'd consider it an accessibility concern.

DaniGuardiola commented:
@tyxla uhhh, it seems like the vertical indicator bug is actually live in trunk and it's my bad. This PR fixes it so we could just finish it up and merge, but if a release is gonna be cut soon I could put up a hotfix. It's quite a trivial fix.

Sorry folks! 😅

tyxla commented:
Can we keep the indicator color changes for another PR? They seem mainly unrelated to this one. 

It's a similar story to the vertical indicator trunk bugs you mentioned. Can they be fixed separately while taking more time to review and test all affected usages here thoroughly? 

In general, I'd prefer that we keep PRs as minimal and atomic as possible. With the initial intention of the PR and all the different types of changes involved, I'm not even sure what to review / test anymore. The list of areas to test was already huge without looking at several different aspects to test.

What do you think?

DaniGuardiola commented:
@tyxla sure, I'll move the indicator color and the bug fix to separate PRs. That accounts for about 3 less lines of changes, everything else is still related to the scrolling/overflow implementation though, so still relevant for a review.

tyxla commented:
> @tyxla sure, I'll move the indicator color and the bug fix to separate PRs. That accounts for about 3 less lines of changes, everything else is still related to the scrolling/overflow implementation though, so still relevant for a review.

Thanks. TO be clear, it's not about the number of lines, but about the impact these lines have. It's one thing to test sizing and overflow behavior (as the PR claims), and it's different when we also need to be testing color changes, and/or some other bug fixes. This is especially valid when there are 10+ distinct places to test.

jameskoster commented:
FWIW, technically the light gray background applied to active vertical tabs does not have sufficient contrast to meet accessibility requirements, so we need to revisit that detail regardless. Using the light blue _for now_ seems like a step in the right direction to me, if we want to fix that detail in this PR.

To ensure a11y needs are met, we'll need to either bump up the contrast significantly, or consider an additional style treatment like reserving bold text for the active tab. That's a detail to discuss in a separate issue/PR imo.

DaniGuardiola committed:
Merge branch 'trunk' of https://github.com/WordPress/gutenberg into feat/tablist-sizing-and-overflow-improvements

DaniGuardiola committed:
Revert vertical indicator changes and some indicator patterns/media tabs styles

DaniGuardiola committed:
Revert vertical indicator bug fix

DaniGuardiola committed:
Add changelog entry

DaniGuardiola commented:
Reverted vertical indicator bug fix and style changes, as well as most of the inserter patterns/media vertical tabs styles. I also added a changelog entry.

Should be ready for review!

DaniGuardiola commented:
Reverted changes submitted as two separate PRs:

- https://github.com/WordPress/gutenberg/pull/65385
- https://github.com/WordPress/gutenberg/pull/65387

tyxla commented on packages/block-editor/src/components/inserter/style.scss:
```diff
@@ -222,73 +222,29 @@ $block-inserter-tabs-height: 44px;
-	flex-direction: column;
-	border: none;
+	&::before {
+		background-color: var(--wp-admin-theme-color) !important;
```
I wish we could avoid the `!important` in general

DaniGuardiola replied:
Yeah... This is emotion getting in the way again. Hopefully when we look into replacing it with something else we can figure out the precedence. Anyway, this style override will be gone for good after https://github.com/WordPress/gutenberg/pull/65387 is done.

tyxla commented on packages/block-editor/src/components/inserter/style.scss:
```diff
@@ -222,73 +222,29 @@ $block-inserter-tabs-height: 44px;
+	}
+
 	margin-bottom: $grid-unit-10;
 	// Push the listitem wrapping the "explore" button to the bottom of the panel.
+	// TODO: is this outdated? The selected element doesn't seem to exist.
 	div[role="listitem"]:last-child {
 		margin-top: auto;
 	}
```
Let's remove this, it's unnecessary since #60970.

DaniGuardiola replied:
Done

tyxla reviewed (commented):
This works well in my testing across the editor and Storybook, nice work @DaniGuardiola!

Could I defer to someone else from @WordPress/gutenberg-components for code review? Also do we need a final look from design?

DaniGuardiola committed:
Remove outdated style.

DaniGuardiola commented:
> Also do we need a final look from design?

I don't believe so, there haven't been any relevant changes since the last design-related discussion resolved.

ciampo commented on packages/components/src/tabs/styles.ts:
```diff
@@ -71,8 +103,7 @@ export const TabListWrapper = styled.div`
 		}
 	}
-	&[aria-orientation='vertical']::after {
-		z-index: -1;
```
Is the `z-index` not necessary anymore because of the switch from `after` to `before`?

DaniGuardiola replied:
correct

ciampo commented on packages/components/src/tabs/tablist.tsx:
```diff
@@ -20,32 +21,46 @@ import type { WordPressComponentProps } from '../context';
+	const selectOnMove = useStoreState( context?.store, 'selectOnMove' );
 
-	if ( ! context || ! tabStoreState ) {
+	if ( ! context || ! context.store ) {
```
Nit

```suggestion
	if ( ! context?.store ) {
```

DaniGuardiola replied:
Done

ciampo commented on packages/components/src/tabs/use-track-overflow.ts:
```diff
@@ -0,0 +1,65 @@
+		} );
+		setObserver( newObserver );
+
+		return () => newObserver?.disconnect();
```
Is the optional chaining necessary? `newObserver` should be always defined

DaniGuardiola replied:
good point, done

ciampo commented on packages/components/src/tabs/use-track-overflow.ts:
```diff
@@ -0,0 +1,65 @@
+import { useState, useEffect } from '@wordpress/element';
+import { useEvent } from '@wordpress/compose';
+
+export function useTrackOverflow(
```
It would be good to add a comment of what this hook does for future references.

DaniGuardiola replied:
done

ciampo commented on packages/components/src/tabs/use-track-overflow.ts:
```diff
@@ -0,0 +1,65 @@
+		}
+		const newObserver = new IntersectionObserver( callback, {
+			root: parent,
+			threshold: 0.9,
```
What are the reasons for not picking `1` as the threshold?

DaniGuardiola replied:
I've seen it fail in some odd cases, so this is just to be safe. It is enough of a margin not to be noticeable while ensuring it doesn't trigger accidentally. It's probably a rounding error by the browser.

ciampo commented on packages/components/src/tabs/styles.ts:
```diff
@@ -87,13 +118,13 @@ export const TabListWrapper = styled.div`
 		align-items: center;
 		position: relative;
 		border-radius: 0;
-		min-height: ${ space(
-			12
-		) }; // Avoid fixed height to allow for long strings that go in multiple lines.
```
What's the reason for removing the `min-height`? 

Also, without it, tabs in storybook have a height of `47.98` px instead of `48px`

DaniGuardiola replied:
Hmm okay, don't remember the exact reason but I'll just re-add it, it wasn't critical.

DaniGuardiola replied:
Re-added but using `height` since there is no wrapping behavior anymore, so the height is always gonna be this fixed value.

ciampo commented on packages/components/src/tabs/styles.ts:
```diff
@@ -123,7 +153,7 @@ export const Tab = styled( Ariakit.Tab )`
 
 		// Focus.
-		&::before {
+		&::after {
```
Any particular reason why we swapped `before` and `after`?

DaniGuardiola replied:
Yes, it was messing with the styles in the inserter patterns/media vertical tabs. The z-index was a hack in the first place since we were using `::after` for something that should have a natural position _before_ the items that show in front.

DaniGuardiola replied:
You can witness this problem yourself by testing this PR: https://github.com/WordPress/gutenberg/pull/65387

Style overrides are gone there and you can see there's no indicator, however if you swap before and after (remember the `transitionend` event), and remove the z-index you'll see it again.

DaniGuardiola replied:
One final note for context is that I only picked `::after` for the indicator because `::before` was taken for the focus stuff to make the most minimal change. There was no other reason and I disliked the z-index from the start, knew something like this was bound to happen :P

ciampo commented on packages/components/src/tabs/stories/index.story.tsx:
Not exactly sure why, but Storybook scrolls the Docs page down on load

https://github.com/user-attachments/assets/60972423-dcbd-4ce9-a0c0-a719003192a1

DaniGuardiola replied:
Ah, good catch. I think I know what it might be. I'll try to fix tomorrow.

DaniGuardiola replied:
actually, just fixed it, it was easy

ciampo commented on packages/components/src/utils/element-rect.ts:
```diff
@@ -79,9 +80,11 @@ export function getElementOffsetRect(
-		element.offsetParent?.getBoundingClientRect() ??
-		NULL_ELEMENT_OFFSET_RECT;
+		offsetParent?.getBoundingClientRect() ?? NULL_ELEMENT_OFFSET_RECT;
+	const offsetParentScrollX = offsetParent?.scrollLeft ?? 0;
```
TIL that `scrollLeft` becomes negative when scrolling with RTL writing direction

DaniGuardiola replied:
Haha, yep. Took me by surprise.

ciampo commented on packages/components/src/tabs/tablist.tsx:
```diff
@@ -64,18 +79,19 @@ export const TabList = forwardRef<
 				/>
 			}
 			onBlur={ onBlur }
+			tabIndex={ -1 }
```
Why are we making the tablist focusable?

DaniGuardiola replied:
This is in place to prevent Firefox from making it focusable by keyboard, which is a behavior that happens automatically when there's overflow in `overflow: auto` containers (see https://github.com/WordPress/gutenberg/pull/64371#issuecomment-2330979556). Tested and it works to prevent that behavior, it's not reachable by keyboard anymore.

ciampo commented on packages/components/src/utils/element-rect.ts:
```diff
@@ -79,9 +80,11 @@ export function getElementOffsetRect(
 	const offsetParentRect =
-		element.offsetParent?.getBoundingClientRect() ??
-		NULL_ELEMENT_OFFSET_RECT;
+		offsetParent?.getBoundingClientRect() ?? NULL_ELEMENT_OFFSET_RECT;
```
I wonder if we can improve runtime performance by "caching" offset parents and scroll positions, and only update them as necessaily

DaniGuardiola replied:
This computation only happens on target changes or resizes. Caching here would defeat the purpose since every time we run the computation we have good reason to think it has potentially changed, and we can't know it has not without accessing this information in the first place.

I'd be surprised if this had any performance issues, the computation rarely runs (on tab change or when the active tab changes size which is now almost impossible since it doesn't even wrap anymore.

DaniGuardiola committed:
Address feedback

DaniGuardiola committed:
Fix scroll bug

DaniGuardiola commented:
@ciampo thanks for the review, I addressed all feedback. Unrelated tests seem to be breaking, I don't know why. Can you take a quick look in case I missed something silly?

ciampo reviewed (commented):
I tried setting as the initially selected tab a tab that would normally be hidden, but the component didn't scroll it into view

```diff
diff --git a/packages/components/src/tabs/stories/index.story.tsx b/packages/components/src/tabs/stories/index.story.tsx
index 3ab1a3c61e..c493fbd85c 100644
--- a/packages/components/src/tabs/stories/index.story.tsx
+++ b/packages/components/src/tabs/stories/index.story.tsx
@@ -122,7 +122,7 @@ export function SizeAndOverflowPlayground() {
 					? 'Remove width: 100% from TabList'
 					: 'Set width: 100% in TabList' }
 			</Button>
-			<Tabs>
+			<Tabs defaultTabId="tab4">
 				<div
 					style={ {
 						width: '20rem',

```

![Screenshot 2024-09-20 at 09 18 34](https://github.com/user-attachments/assets/719a5083-8eb8-47a8-9b1b-f4c5b74fc7cc)


I wonder if https://github.com/WordPress/gutenberg/pull/64371/commits/d3d1d841c00bd549d62fd853b378e8a1714ac1a8 introduced the regression. Is there any part of this new behaviour that we could test with a unit test to avoid such regressions in the future ?

DaniGuardiola commented:
@ciampo good catch. I'm basically bailing out of calling `scrollIntoView` on the initial value because `scrollIntoView` scrolls the whole page and overflow containers in any way necessary to make the element visible, so that's why the Storybook glitchy scroll behavior happened. However, you're right, we still need to make it scroll into view, just only in the overflow container and not in the page as a whole.

Already got a solution in mind that will only scroll the overflow parent.

Re:tests, I don't really think they'll be necessary as this "regression" is just me missing some cases that you correctly caught, after this it seems unlikely that they'd be helpful. That said, if we were to add tests, I'm not sure what those would look like. Open to suggestions if you still see the value of tests here.

DaniGuardiola committed:
Improve automatic tab scrolling behavior.

DaniGuardiola committed:
Merge branch 'trunk' of https://github.com/WordPress/gutenberg into feat/tablist-sizing-and-overflow-improvements

DaniGuardiola committed:
Tweaks to prevent unit test failure and minor cleanup.

DaniGuardiola committed:
Undo unnecessary changes.

DaniGuardiola commented:
@ciampo alright, I added the replacement for `scrollIntoView` and extensively tested it to make sure it works as expected under all reasonable circumstances, let me know if you can manage to break this one 😉

ciampo commented on packages/components/src/tabs/stories/index.story.tsx:
```diff
@@ -70,6 +70,109 @@ const Template: StoryFn< typeof Tabs > = ( props ) => {
 
 export const Default = Template.bind( {} );
 
+export function SizeAndOverflowPlayground() {
```
Let's tweak this new story so that Storybook controls work as expected

```diff
diff --git a/packages/components/src/tabs/stories/index.story.tsx b/packages/components/src/tabs/stories/index.story.tsx
index d64cc1f553..0f7e0d2c6a 100644
--- a/packages/components/src/tabs/stories/index.story.tsx
+++ b/packages/components/src/tabs/stories/index.story.tsx
@@ -70,7 +70,7 @@ const Template: StoryFn< typeof Tabs > = ( props ) => {
 
 export const Default = Template.bind( {} );
 
-export function SizeAndOverflowPlayground() {
+export const SizeAndOverflowPlayground: StoryFn< typeof Tabs > = ( props ) => {
 	const [ fullWidth, setFullWidth ] = useState( false );
 	return (
 		<div>
@@ -123,7 +123,7 @@ export function SizeAndOverflowPlayground() {
 					? 'Remove width: 100% from TabList'
 					: 'Set width: 100% in TabList' }
 			</Button>
-			<Tabs defaultTabId="tab4">
+			<Tabs { ...props }>
 				<div
 					style={ {
 						width: '20rem',
@@ -171,7 +171,10 @@ export function SizeAndOverflowPlayground() {
 			</Tabs>
 		</div>
 	);
-}
+};
+SizeAndOverflowPlayground.args = {
+	defaultTabId: 'tab4',
+};
 
 const VerticalTemplate: StoryFn< typeof Tabs > = ( props ) => {
 	return (

```

DaniGuardiola replied:
done

ciampo reviewed (commented):
The bug that I reported before seems to be fixed, but I found new glitches during my testing:

I was testing `Tabs` in the Site Editor's Stylebook. I manually made some tab labels longer to cause scrolling, and noticed that, while moving the selected tab with arrow keys, the scrollbar flashes briefly before disappearing. (tested on latest chrome, MacOS Sequoia, with external mouse)

https://github.com/user-attachments/assets/7b2fb308-ab62-4bdd-b0e2-01cbbe967022

I also noticed some unexpected behaviour when there are only two items + scrolling. The items are not always fully revealed when receiving focus, and I also wasn't able to scroll with the mouse by dragging the scrollbar.

https://github.com/user-attachments/assets/24dd06e3-b4da-4422-9071-0bf5386acfad

The scrolling behaviour can be also reproduced without editing the storybook example:

https://github.com/user-attachments/assets/115da547-7bb9-4ace-a454-75ff9431629f

The same is probably happening when `selectOnMove={ false }`. When I move focus, sometimes the tablist scrolls back to the selected tab, instead of showing the focused tab


https://github.com/user-attachments/assets/269922e0-ec72-4e13-82b0-aa6bafea7b2b

---

Separately, I reproduced another issue while testing, but I verified that it happens on `trunk` too, and therefore I opened a [separate issue](https://github.com/WordPress/gutenberg/issues/65576). We should still look into addressing this soon, though.

DaniGuardiola committed:
Merge branch 'trunk' of https://github.com/WordPress/gutenberg into feat/tablist-sizing-and-overflow-improvements

DaniGuardiola committed:
Improved story

DaniGuardiola committed:
Fix scroll jumping bug.

DaniGuardiola committed:
Scroll to active tab instead of selected (support `selectOnMove=false`).

DaniGuardiola committed:
Fix minor visual glitch with overflow fade out indicators.

DaniGuardiola committed:
Misc tweaks

DaniGuardiola committed:
Fix.

DaniGuardiola commented:
I think I fixed all issues @ciampo. The first one was caused by an attempt I made to hide the scrollbar when jumping around. Seemed to work when clicking, but it makes it show up for a moment when moving with arrow keys. I just removed that so the scrollbar will just show up. I don't quite love it, but at least it doesn't feel glitchy.

The rest of the bugs should be fixed as well.

DaniGuardiola committed:
Fix changelog

DaniGuardiola committed:
Fix changelog but it's actually true

ciampo commented:
Unfortunately, I can still reproduce the scrollbar flashing in and out of view, both when using keyboard and mouse.

To reproduce it, I load the style book, manually tweak the text of one or more tabs to cause the tablist to overflow, and then interact with the component.

https://github.com/user-attachments/assets/70013459-2146-4d06-aefd-655335b95518


---

Side note: the "close" button in the style book currently sits on top of the tablist and, when the tablist overflows, occludes some of the tablist's content.

![Screenshot 2024-09-25 at 17 38 42](https://github.com/user-attachments/assets/b2c6280e-b357-4b1b-8dd4-09f784640151)

DaniGuardiola commented:
@ciampo huh, must be a different bug than I though then. I reproduced it in the Storybook so that's why I believed it was fixed. I'll look into the specific usage then.

DaniGuardiola committed:
Merge branch 'trunk' of https://github.com/WordPress/gutenberg into feat/tablist-sizing-and-overflow-improvements

DaniGuardiola committed:
Fix changelog

DaniGuardiola commented:
@ciampo after looking at it, I'm pretty sure you accidentally didn't pull or build the latest changes when you left your last comment. Prior to my latest commits, I can repro what you see, but the fix I pushed seems to prevent that correctly. Please do check again and confirm, but in my testing I see no issue.

That said, I didn't think about these tabs potentially overflowing the container. While it seems unlikely to happen when expanded (like in your screenshot), I just realized you can collapse that view to a point where it does happen currently, without changing any labels. I'm pushing some better styles to prevent that in a second.

Let me know if you find any remaining problems :)

DaniGuardiola committed:
Make Story Book tabs nicer.

DaniGuardiola commented:
Here's a demo that shows that the bug is already fixed prior to that comment:

https://github.com/user-attachments/assets/cc8a6c9d-f43f-4000-9281-8099335697e4

DaniGuardiola commented:
And here's a demo of the adjusted layout for Story Book tabs. Notice how the close button does not overlap with the tabs anymore.

https://github.com/user-attachments/assets/d62bebaf-68fc-41b7-85aa-e3a34ec6c3b2

ciampo commented:
> And here's a demo of the adjusted layout for Story Book tabs. Notice how the close button does not overlap with the tabs anymore.

The style book's close button is not overlapping anymore 🎉 

>  I'm pretty sure you accidentally didn't pull or build the latest changes when you left your last comment. 
> ...
> Let me know if you find any remaining problems :)

I'm quite sure I was on the latest commit — but to double-check, I re-pulled the whole PR branch, installed a fresh copy of WP, and tested again. Unfortunately I can still reproduce the same flashing bar:

https://github.com/user-attachments/assets/31a2967b-4aef-4a9a-8e1e-4b34cd57b7da

Also, while testing, I discovered a bit of a weird behaviour while clicking the tab that is right at the edge of the tablist (where the gradient is). Sometimes, after clicking, the indicator disappears — I assume because a new tab is selected, but is not scrolled into view. You can see it happen at the 12s mark, 20s, etc.

https://github.com/user-attachments/assets/22905c0c-d55e-493e-a63d-89255a4dd5ab

ciampo commented:
> 2\. Clicking on the edge doesn't select the tab bug: we tracked it down to an Ariakit bug, see [`TabList`: scrolling active tabs into view breaks selection ariakit/ariakit#4132](https://github.com/ariakit/ariakit/issues/4132)

It looks like [we already have a potential fix](https://github.com/ariakit/ariakit/issues/4132#issuecomment-2379719788) for it 🎉

DaniGuardiola committed:
Temp fix for scrollbar issue in Style Book tabs.

DaniGuardiola committed:
Fix scroll bug and clean up a little.

DaniGuardiola committed:
Simplify and clean up a bit more.

DaniGuardiola commented:
@ciampo the original approach was actually that (to track the selected element). I switched to the active one to cover for `selectOnMove`, however after iterating a bit I found out that we can indeed get away by tracking the selected element, since the active element in all relevant cases always receives focus, and that already causes the container to scroll the element into view. I added the scroll margin back in the styles so that the behavior is similar in both cases.

In my testing I couldn't see anything misbehaving. I tested with different Mac scrollbar settings, different means of interaction, and different combinations of prop values (e.g. `selectOnMove`). Let me know if the fix also works for you 🤞

DaniGuardiola committed:
Merge branch 'trunk' of https://github.com/WordPress/gutenberg into feat/tablist-sizing-and-overflow-improvements

DaniGuardiola committed:
Fix merge issues.

DaniGuardiola committed:
Fix merge issues again.

DaniGuardiola committed:
Make inserter patterns/media changes more minimal

DaniGuardiola commented:
@tyxla fyi I made the inserts > patterns/media tabs changes more minimal in https://github.com/WordPress/gutenberg/pull/64371/commits/e80f76a9be851bf721ca692a59a453a8e8521aa7 since they altered the aspect too much. The vertical tabs styles are being overhauled in https://github.com/WordPress/gutenberg/pull/65387 anyway so this makes for less changes in this PR.

DaniGuardiola committed:
Fix outdated comment

DaniGuardiola committed:
Fix another typo in comment.

DaniGuardiola committed:
Minor cleanup.

ciampo commented on packages/edit-site/src/components/style-book/style.scss:
As mentioned in a comment, let's not forget to follow-up with a refactor of the style book layout that ideally doesn't rely on absolute positioning 👌

ciampo reviewed (approved):
LGTM 🎉 

I did a round of testing on Storybook and in the editor and couldn't find any issues — hopefully I didn't miss anything.

Let's solve the conflicts, merge, and iterate!

Thank you for the patience while working on this PR, @DaniGuardiola 🙏

DaniGuardiola committed:
Merge branch 'trunk' of https://github.com/WordPress/gutenberg into feat/tablist-sizing-and-overflow-improvements

DaniGuardiola committed:
Fix bad automatic merge.

DaniGuardiola committed:
ugh, fix again

stokesman commented:
Dropping a comment here because it seems #65837 is due to these changes. Not saying this PR did anything wrong and perhaps it’s just a follow up needed to fix a detail of `Tabs` usage in the Inserter.
